### DISCUSSION TITLE
how to publish h264 file to rtsp server?

### DISCUSSION THREAD

**Discussion by gadididi** [bluenviron/mediamtx#303] [Questions] [answered]
## Which version are you using?
GST:
v1.0.0
ubuntu 18.04

## Question

how to publish h264 file to the RTSP server?
 this is for mp4 file: gst-launch-1.0 rtspclientsink name=s location=rtsp://localhost:8554/mystream filesrc location=file.mp4 ! qtdemux name=d d.video_0 ! queue ! s.sink_0 d.audio_0 ! queue ! s.sink_1

i tried this: 
gst-launch-1.0 rtspclientsink location=rtsp://localhost:8554/mystream filesrc location=example.h264 ! video/x-h264 ! h264parse ! avdec_h264 ! videoconvert ! videorate ! video/x-raw,framerate=30/1 ! autovideosink

I get on the client-side:
  _Setting pipeline to PAUSED ...
  Pipeline is PREROLLING ...
  Progress: (open) Opening Stream
  Progress: (connect) Connecting to rtsp://localhost:8554/mystream
  Progress: (open) Retrieving server options
  Progress: (open) Opened Stream
  Redistribute latency...
  Redistribute latency...
  Pipeline is PREROLLED ...
  Setting pipeline to PLAYING ...
  New clock: GstSystemClock
  Progress: (request) Sending RECORD request_

and on the server-side:
_2021/03/14 18:28:05 [I] [1/0/0] [client 127.0.0.1:59354] connected (plain)_

what I do wrong?

**Comment by aler9** [comment]
Hello, the problem is that you're converting h264 frames into raw frames (i.e. you're decoding). You don't want to sent to the server the decoded frames, you want to send it the h264 frames, directly from the file, therefore

```
gst-launch-1.0 filesrc location=example.h264 ! video/x-h264 ! h264parse ! rtspclientsink location=rtsp://localhost:8554/mystream
```

**Comment by gadididi** [comment]
> Hello, the problem is that you're converting h264 frames into raw frames (i.e. you're decoding). You don't want to sent to the server the decoded frames, you want to send it the h264 frames, directly from the file, therefore
> 
> ```
> gst-launch-1.0 filesrc location=example.h264 ! video/x-h264 ! h264parse ! rtph264pay ! rtspclientsink location=rtsp://localhost:8554/mystream
> ```

Hi,
thank you, but I got this warning now:

_WARNING: erroneous pipeline: could not link rtph264pay0 to rtspclientsink0_

**Comment by aler9** [comment, marked as answer]
yes, my bad, the right pipeline is

```
gst-launch-1.0 filesrc location=example.h264 ! video/x-h264 ! h264parse ! rtspclientsink location=rtsp://localhost:8554/mystream
```

[a comment by an automated account is not shown]
